coderd/database/queries/provisionerjobs.sql

```diff
@@ -50,42 +50,67 @@ WHERE
 	id = ANY(@ids :: uuid [ ]);
 
 -- name: GetProvisionerJobsByIDsWithQueuePosition :many
-WITH pending_jobs AS (
-    SELECT
-        id, created_at
-    FROM
-        provisioner_jobs
-    WHERE
-        started_at IS NULL
-    AND
-        canceled_at IS NULL
-    AND
-        completed_at IS NULL
-    AND
-        error IS NULL
+WITH filtered_provisioner_jobs AS (
+	-- Step 1: Filter provisioner_jobs and assign an order from upstream system
+	SELECT
+		*,
+		ROW_NUMBER() OVER () AS ordinality -- Track original order
+	FROM
+		provisioner_jobs
+	WHERE
+		id = ANY(@ids :: uuid [ ]) -- Apply filter early to reduce dataset size before expensive JOINs
 ),
-queue_position AS (
-    SELECT
-        id,
-        ROW_NUMBER() OVER (ORDER BY created_at ASC) AS queue_position
-    FROM
-        pending_jobs
+pending_jobs AS (
+	-- Step 2: Extract only pending jobs from the already filtered dataset
+	SELECT *
+	FROM filtered_provisioner_jobs
+	WHERE started_at IS NULL
+		AND canceled_at IS NULL
+		AND completed_at IS NULL
+		AND error IS NULL
 ),
-queue_size AS (
-	SELECT COUNT(*) AS count FROM pending_jobs
+ranked_jobs AS (
+	-- Step 3: Rank only pending jobs based on provisioner availability
+	SELECT
+		pj.id,
+		pj.created_at,
+		ROW_NUMBER() OVER (PARTITION BY pd.id ORDER BY pj.created_at ASC) AS queue_position,
+		COUNT(*) OVER (PARTITION BY pd.id) AS queue_size
+	FROM
+		pending_jobs pj
+			INNER JOIN provisioner_daemons pd
+					ON provisioner_tagset_contains(pd.tags, pj.tags) -- Join only on the small filtered pending set
+),
+final_jobs AS (
+	-- Step 4: Compute best queue position and max queue size per job
+	SELECT
+		fpj.id,
+		fpj.created_at,
+		COALESCE(MIN(rj.queue_position), 0) :: BIGINT AS queue_position, -- Best queue position across provisioners
+		COALESCE(MAX(rj.queue_size), 0) :: BIGINT AS queue_size, -- Max queue size across provisioners
```

Comment:
I understand this `0`. Should we instead use `-1` to indicate no valid `queue_position` or `queue_size`? 

0 could be misinterpreted as "complete" imo. `-1` is more obviously "not in a queue".